crispr_read_counts/utils.py

```diff
@@ -0,0 +1,66 @@
+import gzip
+import re
+from contextlib import contextmanager
+import os
+
+PLASMID_COUNT_HEADER = re.compile(r'^sgRNA\tgene', flags=re.I)
+DNA_PATTERN = re.compile(r'^[ATGC]+$', flags=re.I)
+SAFE_SEQ_FORMAT = re.compile(r'^F\d+$')
+
+
+def error_msg(msg: str):
+  return f'#------\n# Error: {process_multiple_lines(msg)}\n#------'
+
+
+def warning_msg(msg: str):
+  return f'#------\n# Warning: {process_multiple_lines(msg)}\n#------'
+
+
+def process_multiple_lines(msg):
+  return "\n".join(
+    [
+      line if index == 0 else '# ' + line
+      for index, line in enumerate(msg.split('\n'))
+    ]
+  )
+
+
+@contextmanager
+def open_plain_or_gzipped_file(file: str):
+    if file.endswith('.gz'):
+      f = gzip.open(file, 'rt')
+    else:
+      f = open(file, 'r')
+    try:
+        yield f
+    finally:
+        f.close()
+
+
+def rev_compl(dna: str):
```

Comment:
I'd suggest a faster option (here only supporting uppercase letters, as it is implied to be required by `DNA_PATTERN`, though I don't know whether that applies to all sequences `rev_compl` is used with), with execution time about 19-44% of this implementation depending on the length of the sequence:

```python
dna_complement_tr_table = str.maketrans('ACGT', 'TGCA')

def reverse_complement(seq: str) -> str:
    return seq[::-1].translate(dna_complement_tr_table)
```

The current version is slowed down primarily by the call to `reversed` and the repeated initialisation of the translation table.